jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * (C) Copyright 2018-2018, by Christoph Grüne and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.cycle;
+
+import org.jgrapht.Graph;
+import org.jgrapht.graph.GraphWalk;
+import org.jgrapht.util.TypeUtil;
+
+import java.util.*;
+
+/**
+ * Implementation of an algorithm for the local augmentation problem for the cyclic exchange neighborhood,
```

Comment:
The main problem with this implementation is that it is useable only by people who have read the relevant papers about the CMST problem and have also understood the concept of the Improvement Graph and the concept of “subset-disjoint" cycles in the improvement graph.

All these need to be clearly explained inside the Javadoc. 
 - What is the improvement graph? 
 - What is the "subset-disjoint" cycle? 
 - What are the labels that you providing to the algorithm? 
 - What kind of algorithm has produced these labels?
 - Are then any invariants that must be valid for the input?

The big picture is missing here, which means that it is highly unlikely that someone else will use this implementation.